Add Mbed OS support into CHIP

pan- opened the pull request:
#### Problem
There is no support for Mbed OS in CHIP. 

 #### Summary of Changes

Add Mbed OS support to CHIP. 

It requires the PR https://github.com/project-chip/connectedhomeip/pull/6352 to be merged and the Mbed OS docker image to be pushed on dockerhub to pass the build workflow.  

Few details about the port: 
- Our main development target is the [PSOC6 WiFi-BT Prototyping kit](https://os.mbed.com/platforms/CY8CPROTO-062-4343W/) from cypress:
- The port only support WiFi and BLE connectivity; not Thread. 
- There is only two examples ported at the moment: Shell and the lock application. More will come as we continue our effort on CHIP. 
- There is good integration with VSCode, build, flash and debug of the examples supported is provided. 

There is few commits in this PR which could be their own PRs. I want guidance on what I should do with them: 
- https://github.com/project-chip/connectedhomeip/commit/503f4a1eba4c4f4d9ec2457058e90ef53d6469f8: Prevent mutex initialization if locking is not supported. 
- https://github.com/project-chip/connectedhomeip/commit/a4d4181fc69f02d7ecc9f1ada4551b1c01ceb703: Avoid use of -Wno-unknown-warning-option in GCC.
- ~https://github.com/project-chip/connectedhomeip/commit/ec3566b4de1cfc1951afcc157bd89af8e8f71262: Update the Android workflow to archive the APK created by the build.~
- https://github.com/project-chip/connectedhomeip/commit/b693df4bb6f03269c3058a0d926e875122c30d39: Small fix to include C99 format macro constant in some files using them.  
- https://github.com/project-chip/connectedhomeip/commit/726ded34076f7c7ddf03d33937c1f47e04f921e0: Compilation fix for LwIP 2.

In term of future work here is what the Mbed OS team supporting CHIP will work on next: 
- Unit test support: We're adding Mbed OS support in the `test_driver`
- Documentation: This submission is scarce in term of documentation for Mbed OS users we would like to provide a guide that explain how to build a CHIP application with Mbed OS. 
- Example support: We want to support more examples starting with the lightning app then more will follow. 
- Integration testing: It isn't great that no on-device test is run. We've read [CHIP on-device testing](https://github.com/project-chip/connectedhomeip/blob/master/docs/api/device_runner.md) document but it seems there isn't a lot of movement around it at the moment. We would like to contribute in that area.

Damian-Nordic commented on .github/workflows/examples-mbed.yaml:
```diff
@@ -0,0 +1,78 @@
+                  fetch-depth: 0
+                  submodules: true
+
+            - name: Build example Shell App
```
It doesn't have to be Shell App

pan- replied:
Fixed

Damian-Nordic commented on config/mbed/CMakeLists.txt:
```diff
@@ -0,0 +1,279 @@
+    endif()
+endmacro()
+
+macro(chip_gn_arg_flags ARG CFLAGS)
```
I can see it's copied from the nrfconnect code, but note that we've lately slightly changed this approach to make sure that GN args are properly escaped. I suggest you verify that in your solution you can pass flags containing `"` to GN without a build failure.

pan- replied:
Thanks Damian, indeed we borrowed a lot from nrf connect as it uses CMake too. I will update the CMakeLists.txt with the changes present there.

pan- replied:
We actually escape the in `mbed-util.cmake` in the function `mbed_get_compile_definitions`

Damian-Nordic commented on config/mbed/CMakeLists.txt:
```diff
@@ -0,0 +1,279 @@
+chip_gn_arg_bool  ("chip_with_platform_mbedtls"           CONFIG_CHIP_WITH_EXTERNAL_MBEDTLS)
+chip_gn_arg_bool  ("chip_bypass_rendezvous"               CONFIG_CHIP_BYPASS_RENDEZVOUS)
+
+if (BOARD STREQUAL "native_posix")
```
Are those ifs valid for mbed-os as well or can be removed?

pan- replied:
Not applicable at the moment, I will remove and reintroduce at the right time.

pan- replied:
Removed

Damian-Nordic commented on config/mbed/CMakeLists.txt:
```diff
@@ -0,0 +1,279 @@
+    ${CHIP_ROOT}/src/lib/core
+    ${CHIP_ROOT}/third_party/nlassert/repo/include
+    ${CMAKE_CURRENT_BINARY_DIR}/gen/include
+    ${CMAKE_CURRENT_BINARY_DIR}/gen/third_party/connectedhomeip/src/app/include
```
The list of needed include paths has recently been reduced on master. Apply the changes to your patch as well.

pan- replied:
Updated

Damian-Nordic commented on config/mbed/chip-gn/args.gni:
```diff
@@ -0,0 +1,39 @@
+custom_toolchain = "//toolchain:mbed"
+mbedtls_target = "//mbedtls:mbedtls"
+
+treat_warnings_as_errors = false
```
Is it necessary? I think if the problem is with mbedOS headers they can be included using `-isystem` instead `-I` and the rest should compile without warnings, right?

mspang replied:
> #### Problem
> There is no support for Mbed OS in CHIP.
> 
> #### Summary of Changes
> Add Mbed OS support to CHIP.
> 
> It requires the PR #6352 to be merged and the Mbed OS docker image to be pushed on dockerhub to pass the build workflow.
> 
> Few details about the port:
> 
> * Our main development target is the [PSOC6 WiFi-BT Prototyping kit](https://os.mbed.com/platforms/CY8CPROTO-062-4343W/) from cypress:
> * The port only support WiFi and BLE connectivity; not Thread.
> * There is only two examples ported at the moment: Shell and the lock application. More will come as we continue our effort on CHIP.
> * There is good integration with VSCode, build, flash and debug of the examples supported is provided.
> 
> There is few commits in this PR which could be their own PRs. I want guidance on what I should do with them:
> 
> * [503f4a1](https://github.com/project-chip/connectedhomeip/commit/503f4a1eba4c4f4d9ec2457058e90ef53d6469f8): Prevent mutex initialization if locking is not supported.
> * [a4d4181](https://github.com/project-chip/connectedhomeip/commit/a4d4181fc69f02d7ecc9f1ada4551b1c01ceb703): Avoid use of -Wno-unknown-warning-option in GCC.
> * ~[ec3566b](https://github.com/project-chip/connectedhomeip/commit/ec3566b4de1cfc1951afcc157bd89af8e8f71262): Update the Android workflow to archive the APK created by the build.~
> * [b693df4](https://github.com/project-chip/connectedhomeip/commit/b693df4bb6f03269c3058a0d926e875122c30d39): Small fix to include C99 format macro constant in some files using them.
> * [726ded3](https://github.com/project-chip/connectedhomeip/commit/726ded34076f7c7ddf03d33937c1f47e04f921e0): Compilation fix for LwIP 2.
> 
> In term of future work here is what the Mbed OS team supporting CHIP will work on next:
> 
> * Unit test support: We're adding Mbed OS support in the `test_driver`
> * Documentation: This submission is scarce in term of documentation for Mbed OS users we would like to provide a guide that explain how to build a CHIP application with Mbed OS.
> * Example support: We want to support more examples starting with the lightning app then more will follow.
> * Integration testing: It isn't great that no on-device test is run. We've read [CHIP on-device testing](https://github.com/project-chip/connectedhomeip/blob/master/docs/api/device_runner.md) document but it seems there isn't a lot of movement around it at the moment. We would like to contribute in that area.

@rgoliver

pan- replied:
Fixed, when the CHIP library is compiled `-isystem` is used. 
Other warnings have been removed (sometime with Pragma)

Damian-Nordic reviewed (commented):
Reviewed build files so far.

mspang commented on scripts/requirements.txt:
```diff
@@ -130,39 +150,35 @@ pygments==2.8.1
     #   ipython
 pykwalify==1.8.0
     # via west
-pyobjc-core==7.1 ; sys_platform == "darwin"
```
Can you add these back in?

Managing this with multiple platforms is a bit tricky...

pan- replied:
Fixed. 
I missed that, would it be desirable run a multi-platform pip compile command like https://github.com/peterdemin/pip-compile-multi ?

mspang replied:
Maybe, does that solve the platform issues as well? It looks like it just allows a form of inheritance which I think doesn't quite solve the issues with platform specific packages.

mspang commented on src/inet/BUILD.gn:
```diff
@@ -65,6 +65,28 @@ source_set("inet_config_header") {
 }
 
+if (chip_device_platform == "mbed") {
+  source_set("mbed_bsd_socket") {
```
Can you move this to a new file `src/platform/mbed/BUILD.gn`

In fact you should be able to replace this conditional with

  `assert(chip_device_platform == "mbed")`

at the top of `src/platform/mbed/BUILD.gn` if you do that.

pan- replied:
Done

mspang commented on src/platform/mbed/BLEManagerImpl.cpp:
```diff
@@ -0,0 +1,1070 @@
+mbed::DigitalOut led3(LED3, 1);
+#endif
+
+struct ConnectionInfo
```
nit: struct -> class ?

pan- replied:
Changed. 
I haven't found anything in the style guide about this. Could be worth adding a note, this is a pattern commonly used in Mbed OS cpp files.

mspang replied:
Nothing in the CHIP style guide it just struck me as odd that a struct has only private data members. What's the difference in usage between struct and class in Mbed style?

pan- replied:
For header files, struct are for PODs or plain interfaces class (no member and only pure virtual function) as it makes sense in both of these case to have everything public. Class is used for everything else.

For classes internal to a `cpp` file which often are _companion_ classes it makes sense to make everything public by default. 

We're fine with everything and consistency is what matter the most. So we will make sure to use `class` for everything except PODs in this codebase.

mspang commented on config/mbed/CMakeLists.txt:
```diff
@@ -0,0 +1,279 @@
+)
+
+list(APPEND CHIP_DEFINES 
+    __LINUX_ERRNO_EXTENSIONS__=1
```
What does this do?

Is errno task local in Mbed ?

pan- replied:
We need that define to access all the error codes required by our BSD socket implementation.

mspang replied:
Is errno task local in Mbed ?

pan- replied:
I'm not sure to understand what you mean by _task_. The `errno` variable (not really a variable by default with newlib) should be global to the application and there is mechanism in place in newlib and Mbed to handle concurrent access.

mspang commented:
It seems this adds support for Mbed by adding adapters for various POSIX (sockets) and even Linux-specific APIs (eventfds).

Is the aim to just keep using an adapter to POSIX & Linux APIs going forward?

pan- commented on examples/lock-app/mbed/main/AppTask.cpp:
```diff
@@ -0,0 +1,416 @@
+    BoltLockMgr().SetCallbacks(ActionInitiated, ActionCompleted);
+
+    // Start BLE advertising if needed
+    if (!CHIP_DEVICE_CONFIG_CHIPOBLE_ENABLE_ADVERTISING_AUTOSTART)
```
We are quite limited on button at the moment. On our reference board more capacitive buttons are available and we're adding support for them. Is it something that must be fixed in this PR or can it wait a subsequent PR ?

Damian-Nordic replied:
No, I just letting you know, but I'm fine with fixing things later :)

pan- replied:
Tracked in https://github.com/ARMmbed/connectedhomeip/issues/166 .

pan- commented on examples/lock-app/mbed/main/AppTask.cpp:
```diff
@@ -0,0 +1,416 @@
+        if (nowUS > nextChangeTimeUS)
+        {
+            // TODO:
+            // PublishService();
```
Removed

pan- commented on examples/lock-app/mbed/main/include/CHIPProjectConfig.h:
```diff
@@ -0,0 +1,38 @@
+#define CHIP_DEVICE_CONFIG_ENABLE_WIFI_AP 1
+
+// Use a default pairing code if one hasn't been provisioned in flash.
+#define CHIP_DEVICE_CONFIG_USE_TEST_SETUP_PIN_CODE 12345678
```
Fixed

pan- commented on examples/shell/mbed/CHIPProjectConfig.h:
```diff
@@ -0,0 +1,33 @@
+#define CHIP_DEVICE_CONFIG_ENABLE_WIFI_AP 0
+
+// Use a default pairing code if one hasn't been provisioned in flash.
+#define CHIP_DEVICE_CONFIG_USE_TEST_SETUP_PIN_CODE 12345678
```
Fixed

pan- commented on examples/shell/mbed/cmd_mbed_utils.cpp:
```diff
@@ -0,0 +1,1020 @@
+static SecurePairingUsingTestSecret gTestPairing;
+static AdminPairingTable gAdmin;
+
+int cmd_server_on(int argc, char ** argv)
```
For sure they can but not in this PR.

pan- commented on scripts/setup/raspberry_pi/setup_network_env.sh:
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env bash
```
Renamed

pan- commented on src/inet/BUILD.gn:
```diff
@@ -65,6 +65,28 @@ source_set("inet_config_header") {
+    public_configs = [ "${chip_root}/src:includes" ]
+
+    sources = [
+      "${chip_root}/src/platform/mbed/net/src/BSDSocket.cpp",
```
These files will be upstreamed with proper tests to Mbed OS.

pan- commented on src/platform/mbed/BLEManagerImpl.cpp:
```diff
@@ -0,0 +1,1070 @@
+namespace Internal {
+
+namespace {
+const UUID ShortUUID_CHIPoBLEService(0xFEAF);
```
Replaced

pan- commented:
@Damian-Nordic @mspang I think I addressed all your comments. Fixes are in new commit that I can rebase once everything has been reviewed.

> Is the aim to just keep using an adapter to POSIX & Linux APIs going forward?

As long as CHIP doesn't expose a platform abstraction for networking, we would _prefer_ keep it that way. We could have chosen to go the LWIP way but this would cut part of our user base out of CHIP if they use an off-board stack and it could cause other issues with part of the user application that wants to use Mbed OS API for networking instead of CHIP API.

The plan is to upstream the BSD socket layer we created to Mbed OS in a near future so this won't stay forever in that repository.

mspang commented on src/platform/mbed/BUILD.gn:
```diff
@@ -0,0 +1,24 @@
+    public_configs = [ "${chip_root}/src:includes" ]
+
+    sources = [
+        "${chip_root}/src/platform/mbed/net/src/BSDSocket.cpp",
```
nit: just "src/BSDSocket.cpp"

pan- replied:
Thanks, I made the change and rebased against `master` as CI was failing.

pan- commented:
@mspang Is there anything else I can do to move this forward ?

Damian-Nordic commented on src/platform/mbed/ConfigurationManagerImpl.h:
```diff
@@ -0,0 +1,103 @@
+#include <platform/mbed/MbedConfig.h>
+
+// Forward declare test method that run the tests.
+int cmd_device_test_config(int argc, char ** argv);
```
Why doesn't the command call `ConfigurationMgr().RunUnitTests` which is public and wouldn't need such "friend" declarations?

pan- replied:
Because we want this to be available even if `NDEBUG` is defined. That's not possible today: 
https://github.com/project-chip/connectedhomeip/blob/17be4afe670126e94282bcf14a57aa3984e36cf4/src/include/platform/ConfigurationManager.h#L570-L571 .

Damian-Nordic commented on src/system/SystemConfig.h:
```diff
@@ -682,6 +682,20 @@ struct LwIPEvent;
 #endif
 #endif // CHIP_SYSTEM_CONFIG_USE_ZEPHYR_NET_IF
 
+#ifndef CHIP_SYSTEM_CONFIG_USE_MBED_NET_IF
```
Is it used?

pan- replied:
Not anymore, initially we didn't had the bsd netif emulation in place and used special case for mbed.

pan- replied:
I removed it.

andy31415 commented on build/config/compiler/BUILD.gn:
```diff
@@ -158,6 +158,8 @@ config("disabled_warnings") {
       "-Wno-cast-function-type",
     ]
+  } else {
+    cflags += [ "-Wno-unknown-warning-option" ]
```
This seems a larger global change.Is this required?

Since the above cflags already sets it, is this setting it again?

pan- replied:
It is not needed and I removed it local. I wanted to push but after rebase the lock application is not working anymore. I'm investigating and will push in this PR when fixed.

pan- replied:
@andy31415 I removed that change and re-enable all warnings that were disabled for Mbed OS builds.

pan- commented:
I believe the PR address all the comments made. 
However the lock app continue to fails though since the last rebase. 
https://github.com/project-chip/connectedhomeip/pull/6842 helped but now I see the error reported in: https://github.com/project-chip/connectedhomeip/issues/6867

mspang reviewed (commented):
> @mspang Is there anything else I can do to move this forward ?

What's the ETA for landing the POSIX adaptation in your SDK?

pan- commented on .vscode/c_cpp_properties.json:
```diff
@@ -82,6 +82,23 @@
+            "cStandard": "c11",
+            "cppStandard": "gnu++14",
+            "intelliSenseMode": "gcc-arm",
+            "compilerPath": "/opt/mbed-os-toolchain/gcc-arm-none-eabi-9-2019-q4-major/bin/arm-none-eabi-gcc",
```
Yes it is: 

https://github.com/project-chip/connectedhomeip/blob/f6f2bfa768841a0df9775701368d2169bda4e395/integrations/docker/images/chip-build-mbed-os/Dockerfile#L37-L50

mspang replied:
Can you dedupe this please, we already had a copy of ARM gcc:

https://github.com/project-chip/connectedhomeip/blob/f6f2bfa768841a0df9775701368d2169bda4e395/integrations/docker/images/chip-build-vscode/Dockerfile#L18

pan- replied:
Sure, I will open a PR to modify the docker image.

pan- commented:
@mspang I don't have a precise picture on when the BSD socket adaptation will land in Mbed OS. My aim is before the end of the summer. 

Is it a concern to have it in the Mbed OS port for now ? We can move it out of CHIP in another repo under the `ARMmbed` organization as it is independent from CHIP.

mspang commented:
Landing this now risks Mbed becoming the last platform using the select() API for I/O polling. Adapting to a partial implementation of the POSIX APIs we were using puts us in an awkward situation; Mbed may be missing something that we need.

+ @kpschoedel to comment on how disruptive merging this now would be to the work on https://github.com/project-chip/connectedhomeip/issues/5556 .

pan- commented:
I'm not sure to understand how the problematic is different from Linux or Zephyr which uses select today. Is there a proposed API for #5556 ?
We can collaborate with the individuals handling this issue to make the necessary changes but that would be good to have some visibility on what the change is and when it would land.

kpschoedel commented:
> Landing this now risks Mbed becoming the last platform using the select() API for I/O polling. Adapting to a partial implementation of the POSIX APIs we were using puts us in an awkward situation; Mbed may be missing something that we need.
> 
> * @kpschoedel to comment on how disruptive merging this now would be to the work on #5556 .

I've only been working with the _POSIX implementation, with the intent to convert the _Zephyr version afterward. This version looks substantially similar, so I don't think a third copy makes much difference in that respect. Practically speaking, the immediate platform work is just shuffling some fragments of `SysUpdate()` and `SysProcess()`.

The first draft API for #5556 was #6561 (see *EndPoint.cpp for examples of use). My current WIP is substantially similar but also encapsulates the file descriptor and active conditions (`mPendingIO`) to reduce duplication. I hope to put up a revised draft with better API comments tomorrow.

One thing I notice in this PR is a change to the functionality of `UDPEndPoint::SendMsg()` that doesn't seem mbed-related.

pan- commented:
Thanks, I will review #6561. 

> One thing I notice in this PR is a change to the functionality of UDPEndPoint::SendMsg() that doesn't seem mbed-related.

It addresses and issue in `UDPEndPoint::SendMsg`. `SendMsg` can be used directly and it will open a new socket if one hasn't been open: 

https://github.com/project-chip/connectedhomeip/blob/f6f2bfa768841a0df9775701368d2169bda4e395/src/inet/UDPEndPoint.cpp#L647-L655

If `select` is already running, this new socket is not part of the file descriptor being monitored. The call to `WakeSelect` will force the system to include that socket in the next `select` call.

mspang commented:
> Thanks, I will review #6561.
> 
> > One thing I notice in this PR is a change to the functionality of UDPEndPoint::SendMsg() that doesn't seem mbed-related.
> 
> It addresses and issue in `UDPEndPoint::SendMsg`. `SendMsg` can be used directly and it will open a new socket if one hasn't been open:
> 
> https://github.com/project-chip/connectedhomeip/blob/f6f2bfa768841a0df9775701368d2169bda4e395/src/inet/UDPEndPoint.cpp#L647-L655
> 
> If `select` is already running, this new socket is not part of the file descriptor being monitored. The call to `WakeSelect` will force the system to include that socket in the next `select` call.

These generic bug fixes or other uncoupled changes should really be separated out into standalone PRs.

mspang commented:
@pan- Your PR was approved, are you able to rebase it?

woody-apple commented:
/rebase

pan- commented:
@mspang I was planning in making a new one since that was was marked at stalled while the BLE flow was broken... 
It is fixed here now and we kept rebasing our port. I can push on this PR the update for sure. However few things have changed since the review: 
- Our posix adaptation layer has been moved in another repository 
- MDNS has been enabled for UDP. 

We also support more examples but I will open new PRs for them.

mspang commented:
> @mspang I was planning in making a new one since that was was marked at stalled while the BLE flow was broken...
> It is fixed here now and we kept rebasing our port. I can push on this PR the update for sure. However few things have changed since the review:
> 
> * Our posix adaptation layer has been moved in another repository
> * MDNS has been enabled for UDP.
> 
> We also support more examples but I will open new PRs for them.

Ok, opening a new PR sounds fine.
